lib/scholar/metrics/regression.ex

```diff
@@ -133,6 +132,150 @@ defmodule Scholar.Metrics.Regression do
     |> Nx.mean()
   end
 
+  @doc """
+  Calculates the mean Tweedie deviance of predictions
+  with respect to targets. Includes the Gaussian, Poisson,
+  Gamma and inverse-Gaussian families as special cases.
+
+  #{~S'''
+  $$d(y,\mu) =
+  \begin{cases}
+  (y-\mu)^2, & \text{for }p=0\\\\
+  2(y \log(y/\mu) + \mu - y), & \text{for }p=1\\\\
+  2(\log(\mu/y) + y/\mu - 1), & \text{for }p=2\\\\
+  2\left(\frac{\max(y,0)^{2-p}}{(1-p)(2-p)}-\frac{y\mu^{1-p}}{1-p}+\frac{\mu^{2-p}}{2-p}\right), & \text{for }p<0 \vee p>2
+  \end{cases}$$
+  '''}
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([1, 1, 1, 1, 1, 2, 2, 1, 3, 1], type: :u32)
+      iex> y_pred = Nx.tensor([2, 2, 1, 1, 2, 2, 2, 1, 3, 1], type: :u32)
+      iex> Scholar.Metrics.Regression.mean_tweedie_deviance(y_true, y_pred, 1)
+      #Nx.Tensor<
+        f32
+        0.18411168456077576
+      >
+  """
+  deftransform mean_tweedie_deviance(y_true, y_pred, power) do
+    message = "Mean Tweedie deviance with power=#{power} can only be used on "
+
+    case check_tweedie_deviance_power(y_true, y_pred, power) |> Nx.to_number() do
+      2 -> raise message <> "strictly positive y_pred."
+      4 -> raise message <> "non-negative y_true and strictly positive y_pred."
+      5 -> raise message <> "strictly positive y_true and strictly positive y_pred."
+      100 -> raise "Something went wrong, branch should never appear."
+      1 -> :ok
+    end
```

Comment:
We cannot use runtime checks, because they would fail when function would be called in another `defn`

Reply:
You can test this by calling `Nx.Defn.jit(&mean_tweedie_deviance/3).(y_true, y_pred, 1)`. You will see that jit returns an expression and you can't convert it to number.

Reply:
Do you mean this:
```
iex> Nx.Defn.jit(&mean_tweedie_deviance/3).(y_true, y_pred, 1)
** (Protocol.UndefinedError) protocol String.Chars not implemented for #Nx.Tensor<
  s64

  Nx.Defn.Expr
  parameter a:2   s64
> of type Nx.Tensor (a struct). This protocol is implemented for the following type(s): Atom, BitString, Complex, Date, DateTime, Float, Hex.Solver.Assignment, Hex.Solver.Constraints.Empty, Hex.Solver.Constraints.Range, Hex.Solver.Constraints.Union, Hex.Solver.Incompatibility, Hex.Solver.PackageRange, Hex.Solver.Term, Integer, List, NaiveDateTime, Time, URI, Version, Version.Requirement
    (elixir 1.14.5) lib/string/chars.ex:3: String.Chars.impl_for!/1
    (elixir 1.14.5) lib/string/chars.ex:22: String.Chars.to_string/1
    (scholar 0.2.1) lib/scholar/metrics/regression.ex:161: Scholar.Metrics.Regression.mean_tweedie_deviance/3
    (nx 0.7.0-dev) lib/nx/defn/compiler.ex:158: Nx.Defn.Compiler.runtime_fun/3
    (nx 0.7.0-dev) lib/nx/defn/evaluator.ex:82: Nx.Defn.Evaluator.precompile/3
    (nx 0.7.0-dev) lib/nx/defn/evaluator.ex:60: Nx.Defn.Evaluator.__compile__/4
    (nx 0.7.0-dev) lib/nx/defn/evaluator.ex:53: Nx.Defn.Evaluator.__jit__/5
    iex:4: (file)
```

I thought that by using the `deftransform`, it could be done. That assumption from reading the docs, was the ideia behing the `check_tweedie_deviance_power` returning a tensor "error value". But from what I understand now, what you mean is that this happens when a `defn` function calls the `mean_tweedie_deviance` defined with `deftransform`.

If so, do you propose any idea/solution for it. In this case, the values need to be checked in order to guarantee that the results "can be calculated" as to have probabilistic meaning.

Reply:
You need to imagine that `defn` is running inside the GPU. Inside the GPU you can't have different shapes and you can't have exceptions. The best you can do is return NaN to signal that something went wrong OR have a separate function, that is a regular `def`, that receives a tensor to check it. In this case, the best is to assume you receive valid inputs and, if not, the value is unspecified.

Reply:
I think having the checks might be useful, at least for sanity check. I've added an option to allow for that check to happen in a `def` function, which by default is false (no check is made). Which means that by default, if the incorrect inputs for the given power value are passed to these functions, there is no guarantee of a result (with some NaNs appearing). Let me know if this can be done like this, or if you have any other concerns.

Reply:
Maybe a better solution then is to have `mean_tweedie_deviance` (no checking) and `mean_tweedie_deviance!` (with checking)?

Reply:
I was following the opts approach, but if possible I agree. In that case would `mean_tweedie_deviance!` be a `def` function?

Reply:
Yeah, it would be a `def`!

Reply:
That would be great, yeah!

Reply:
Ok, added it as well.